### PR TITLE
infer some descriptor_rows from table_config

### PR DESCRIPTION
**Pull request by turbomam**
- #113

### CODE REVIEW

File: schemasheets/schema_exporter.py [hunk 1]
```diff
@@ -26,6 +27,77 @@ def _configuration_has_primary_keys_for(table_config: TableConfig, metatype: str
...
+
+    fields: list[str] = []
+    for attribute in cls.__dict__:
+        if not attribute.startswith('_'):
+            fields.append(attribute)
+
+    return fields
+
+
+def infer_descriptor_rows(table_config: TableConfig) -> List[ROW]:
```
Comment (cmungall) [inline schemasheets/schema_exporter.py:49]: can you add a docstring?

File: schemasheets/schema_exporter.py [hunk 2]
```diff
@@ -26,6 +27,77 @@ def _configuration_has_primary_keys_for(table_config: TableConfig, metatype: str
     return False
 
 
+def get_fields(cls):
```
Comment (cmungall) [inline schemasheets/schema_exporter.py:30]: add typing

File: schemasheets/examine_schemasheet.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+from schemasheets.schemasheet_datamodel import SchemaSheet
```
Comment (cmungall) [inline schemasheets/examine_schemasheet.py:1]: Can you document this?

File: schemasheets/examine_schemasheet.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+
+specification = "../populated_with_generated_spec_subset.tsv"
```
Comment (cmungall) [inline schemasheets/examine_schemasheet.py:3]: avoid relative links. See e.g. https://stackoverflow.com/questions/2817264/how-to-get-the-parent-dir-location

File: pyproject.toml [hunk 5]
```diff
@@ -32,3 +32,4 @@ build-backend = "poetry.core.masonry.api"
 sheets2linkml = "schemasheets.schemamaker:convert"
 linkml2sheets = "schemasheets.schema_exporter:export_schema"
 sheets2project = "schemasheets.sheets_to_project:multigen"
+generate-populate = 'schemasheets.generate_populate:cli'
```
Comment (cmungall) [inline pyproject.toml:35]: remember, these commands may be installed at a system level. `generate-populate` is a very generic script name.

Can you choose something that fits with the existing conventions and is less generic? E.g. linkml2sheets-template?

File: schemasheets/generate_populate.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+        format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
+        filename=log_file,
+    )
+
+
+def tabulate_unique_values(list_):
+    """
+    tabulate the number of appearances of each unique values in a list
+
+    Args:
```
Comment (cmungall) [inline schemasheets/generate_populate.py:82]: See https://berkeleybop.github.io/best_practice/

* Always provide docstrings
     * ReST (reStructuredText) » numpy-style docstrings or google style » nothing

File: schemasheets/generate_populate.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+    for value in unique_values:
+        count = list_.count(value)
+        value_counts[value] = count
+
+    sorted_value_counts = sorted(value_counts.items(), key=lambda x: x[1], reverse=True)
+
+    return sorted_value_counts
+
+
+def discover_source_usage(source_view):
```
Comment (cmungall) [inline schemasheets/generate_populate.py:100]: add typing

File: schemasheets/generate_populate.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+                    discovered_meta_slots.append(cisk)
+
+            cis_annotations_dict = as_dict(cis.annotations)
+            for a in cis_annotations_dict:
+                discovered_annotations.append(a)
+
+    return discovered_meta_slots, discovered_annotations
+
+
+def do_usage_report(style, meta_view, meta_type_names, meta_enum_names, discovered_annotations,
```
Comment (cmungall) [inline schemasheets/generate_populate.py:128]: add typing

File: schemasheets/generate_populate.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+                    discovered_source_slots, logger):
+    logger.warning(style)
```
Comment (cmungall) [inline schemasheets/generate_populate.py:130]: did you mean to leave this in?

File: schemasheets/generate_populate.py [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+from click import command, option, Choice
+from jsonasobj2 import as_dict
+from linkml_runtime import SchemaView
+from schemasheets.conf.configschema import ColumnSettings
+from schemasheets.schema_exporter import SchemaExporter
+from schemasheets.schemasheet_datamodel import TableConfig, ColumnConfig
+import logging
+import pkg_resources
+
+# todo write tests
```
Comment (cmungall) [inline schemasheets/generate_populate.py:11]: good idea :-)

File: schemasheets/examine_schemasheet.py (hunk 3 above)
Comment (turbomam) [reply schemasheets/examine_schemasheet.py:1]: I'm just going to delete that script. It was for learning by reverse engineering.

File: schemasheets/examine_schemasheet.py (hunk 4 above)
Comment (turbomam) [reply schemasheets/examine_schemasheet.py:3]: thanks for the reminder. solving by deleting the script.

File: schemasheets/examine_schemasheet.py (hunk 3 above)
Comment (turbomam) [reply schemasheets/examine_schemasheet.py:1]: I'm deleting this script.It was for learning by reverse engineering.

File: schemasheets/generate_populate.py (hunk 9 above)
Comment (turbomam) [reply schemasheets/generate_populate.py:130]: No!

File: schemasheets/schema_exporter.py [hunk 11]
```diff
@@ -26,6 +27,77 @@ def _configuration_has_primary_keys_for(table_config: TableConfig, metatype: str
...
+    cs_fields = get_fields(ColumnSettings)
+    cs_fields.sort()
+
+    handled_attributes = ["header", ] + cs_fields
+
+    descriptor_rows: list[dict[str, str]] = []
+
+    for ha in handled_attributes:
+        index = 0
+        temp_dict = {}
```
[the remark this answers is not in the document]
Comment (turbomam) [reply schemasheets/schema_exporter.py:59]: I fixed this one. I'll add an issue to find and address more

[the remark this answers is not in the document]
Comment (turbomam) [reply schemasheets/schema_exporter.py:59]: https://github.com/linkml/schemasheets/issues/118

File: schemasheets/generate_populate.py (hunk 6 above)
Comment (turbomam) [reply schemasheets/generate_populate.py:82]: Added or changed docstrings to ReST style

**Comment by turbomam** [comment]
https://github.com/linkml/schemasheets/issues/119

**Pull request merged**